pulsar-broker/src/test/java/org/apache/pulsar/client/api/UnloadSubscriptionTest.java

```diff
@@ -0,0 +1,243 @@
+    }
+
+    @Override
+    protected void doInitConf() throws Exception {
+        super.doInitConf();
+        conf.setSystemTopicEnabled(false);
+        conf.setTransactionCoordinatorEnabled(false);
+    }
+
+    @AfterClass(alwaysRun = true)
+    @Override
+    protected void cleanup() throws Exception {
+        super.internalCleanup();
+    }
+
+    @DataProvider(name = "unloadCases")
+    public Object[][] unloadCases (){
+        // [msgCount, enabledBatch, maxMsgPerBatch, subType, ackMsgCount]
+        return new Object[][]{
+                {100, false, 1, Exclusive, 0},
+                {100, false, 1, Failover, 0},
+                {100, false, 1, Shared, 0},
+                {100, false, 1, Key_Shared, 0},
+                {100, true, 5, Exclusive, 0},
+                {100, true, 5, Failover, 0},
+                {100, true, 5, Shared, 0},
+                {100, true, 5, Key_Shared, 0},
+                {100, false, 1, Exclusive, 50},
+                {100, false, 1, Failover, 50},
+                {100, false, 1, Shared, 50},
+                {100, false, 1, Key_Shared, 50},
+                {100, true, 5, Exclusive, 50},
+                {100, true, 5, Failover, 50},
+                {100, true, 5, Shared, 50},
+                {100, true, 5, Key_Shared, 50},
+        };
+    }
+
+    @Test(dataProvider = "unloadCases", invocationCount = 50)
+    public void testSingleConsumer(int msgCount, boolean enabledBatch, int maxMsgPerBatch, SubscriptionType subType,
+                                   int ackMsgCount) throws Exception {
+        final String topicName = "persistent://my-property/my-ns/tp-" + UUID.randomUUID().toString();
+        final String subName = "sub";
+        ListenerConsumer listenerConsumer = createListener(topicName, subName, subType);
+        ProducerAndMessageIds producerAndMessageIds =
+                createProducerAndSendMessages(topicName, msgCount, enabledBatch, maxMsgPerBatch);
+        log.info("send message-ids:{}-{}", producerAndMessageIds.messageIds.size(),
+                toString(producerAndMessageIds.messageIds));
+        Awaitility.await().untilAsserted(() -> {
+            Set<String> allMessages = listenerConsumer.listener.messageSet;
+            assertEquals(allMessages.size(), msgCount);
+        });
+
+        if (ackMsgCount > 0){
+            List<MessageId> messageIdsToAck = producerAndMessageIds.messageIds.subList(0, ackMsgCount);
+            log.info("ack message-ids: {}", toString(messageIdsToAck));
+            listenerConsumer.consumer.acknowledge(messageIdsToAck);
+        }
+
+        listenerConsumer.listener.messageSet.clear();
+        PersistentTopic persistentTopic = getPersistentTopic(topicName);
+        persistentTopic.unloadSubscription(subName);
+        Awaitility.await().untilAsserted(() -> {
+            Set<String> allMessages = listenerConsumer.listener.messageSet;
+            assertEquals(allMessages.size(), msgCount - ackMsgCount);
```

Comment:
this assert, when consumer util to msgCount - ackMsgCount the return directly, so we haven check the consumer can't receive the message anymore. same as `testMultiConsumer`

Reply:
After covering so many scenarios, it is possible to get duplicate messages. I changed the implementation to receive all messages before the verification